### ISSUE TITLE
Reproducible K-means Clustering Pipeline for Biomarker-Based Phenotyping

### ISSUE THREAD

**Issue by Curiosity-Mars** [Curiosity-Mars/total-testosterone#1]
"""
Clustering pipeline for biomarker data (E-1 outputs).

NOTE:
- Originally written in Japanese; this public version is translated and cleaned in English.
- This script was developed in Google Colab and expects the input Excel file to be accessible
  via a filesystem path (often Google Drive mounted under /content/drive/).

IMPORTANT (for readers):
- If you are NOT using the original Google Drive location, you MUST replace --xlsx_path
  with your own local path (or your own mounted Drive path in Colab).
"""

import os
import argparse
import numpy as np
import pandas as pd

from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn.metrics import (
    silhouette_score,
    calinski_harabasz_score,
    davies_bouldin_score,
)

EXPECTED_COLS = [
    "age", "height", "BW", "BMI", "Hb", "Hct", "Alb", "AST", "ALT", "ALP", "Cre",
    "Glu", "TG", "HDL-C", "LDL-C", "LH", "FSH", "CRP", "TT"
]

def load_and_validate(xlsx_path: str) -> pd.DataFrame:
    if not os.path.exists(xlsx_path):
        raise FileNotFoundError(f"File not found: {xlsx_path}")

    df = pd.read_excel(xlsx_path)
    missing = [c for c in EXPECTED_COLS if c not in df.columns]
    if missing:
        raise ValueError(f"Missing columns in input file: {missing}")

    return df

def preprocess(df: pd.DataFrame):
    """
    Mean imputation -> Standardization.
    Returns:
      X_raw (DataFrame): original scale (subset of EXPECTED_COLS)
      X (ndarray): standardized features after mean imputation
      imputer, scaler: fitted transformers
    """
    X_raw = df[EXPECTED_COLS].copy()

    imputer = SimpleImputer(strategy="mean")
    scaler = StandardScaler()

    X_imp = imputer.fit_transform(X_raw)
    X = scaler.fit_transform(X_imp)

    return X_raw, X, imputer, scaler

def eval_k(X: np.ndarray, k: int, seed: int = 1, n_init: int = 50, sil_sample: int = 2000) -> dict:
    km = KMeans(n_clusters=k, random_state=seed, n_init=n_init)
    labels = km.fit_predict(X)

    # Silhouette can be expensive; optionally sample rows deterministically.
    if X.shape[0] > sil_sample:
        rng = np.random.default_rng(seed)
        idx = rng.choice(X.shape[0], sil_sample, replace=False)
        sil = silhouette_score(X[idx], labels[idx])
    else:
        sil = silhouette_score(X, labels)

    return {
        "k": k,
        "SSE": float(km.inertia_),
        "silhouette": float(sil),
        "CH": float(calinski_harabasz_score(X, labels)),
        "DB": float(davies_bouldin_score(X, labels)),
        "min_cluster_size": int(np.min(np.bincount(labels))),
    }

def tiny_cluster_seed_scan(X: np.ndarray, outdir: str, ks=(3, 4), seeds=range(50), n_init: int = 10):
    rows = []
    for k in ks:
        for seed in seeds:
            km = KMeans(n_clusters=k, random_state=seed, n_init=n_init)
            labels = km.fit_predict(X)
            rows.append({
                "k": int(k),
                "seed": int(seed),
                "min_cluster_size": int(np.min(np.bincount(labels))),
            })

    tiny_raw = pd.DataFrame(rows)
    tiny_raw.to_csv(os.path.join(outdir, "tiny_cluster_seed_scan_raw.csv"), index=False)

    tiny_summary = tiny_raw.groupby("k")["min_cluster_size"].describe()
    tiny_summary.to_csv(os.path.join(outdir, "tiny_cluster_seed_scan_summary.csv"))

    return tiny_raw, tiny_summary

def subgroup_entropy(X_raw: pd.DataFrame, labels: np.ndarray):
    """
    Define subgroup as: TT <= Q25 and CRP >= Q75 (on original scale).
    Compute distribution across clusters and entropy (bits).
    """
    tt_q25 = float(X_raw["TT"].quantile(0.25))
    crp_q75 = float(X_raw["CRP"].quantile(0.75))

    mask = (X_raw["TT"] <= tt_q25) & (X_raw["CRP"] >= crp_q75)
    sub = pd.Series(labels[mask.values]).value_counts().sort_index()

    p = sub / sub.sum()
    entropy_bits = float(-(p * np.log2(p)).sum()) if len(p) else 0.0

    return tt_q25, crp_q75, int(mask.sum()), entropy_bits, sub.to_dict()

def compute_subgroup_split(X_raw: pd.DataFrame, X: np.ndarray, outdir: str, ks=(3, 4), seed: int = 1, n_init: int = 50):
    rows = []
    for k in ks:
        km = KMeans(n_clusters=k, random_state=seed, n_init=n_init)
        labels = km.fit_predict(X)

        tt_thr, crp_thr, nsub, ent, dist = subgroup_entropy(X_raw, labels)
        rows.append({
            "k": int(k),
            "TT_Q25_threshold": float(tt_thr),
            "CRP_Q75_threshold": float(crp_thr),
            "subgroup_n": int(nsub),
            "entropy_bits": float(ent),
            "distribution": dist,
        })

    sub_df = pd.DataFrame(rows)
    sub_df.to_csv(os.path.join(outdir, "subgroup_lowTT_highCRP_split.csv"), index=False)
    return sub_df

def centers_df(X: np.ndarray, scaler: StandardScaler, k: int, seed: int = 1, n_init: int = 50) -> pd.DataFrame:
    km = KMeans(n_clusters=k, random_state=seed, n_init=n_init)
    km.fit(X)

    centers_original = scaler.inverse_transform(km.cluster_centers_)
    dfc = pd.DataFrame(centers_original, columns=EXPECTED_COLS)
    dfc.insert(0, "cluster", range(k))
    return dfc

def main():
    parser = argparse.ArgumentParser(description="K-means evaluation and output generation.")
    parser.add_argument(
        "--xlsx_path",
        type=str,
        default="/content/drive/MyDrive/CommmedpatientTT.xlsx",
        help="Path to the input Excel file. In Colab, this often points to mounted Google Drive."
    )
    parser.add_argument(
        "--outdir",
        type=str,
        default="e1_outputs",
        help="Directory to save output CSV files."
    )
    parser.add_argument(
        "--seed_fixed",
        type=int,
        default=1,
        help="Random seed for fixed-seed evaluations."
    )
    args = parser.parse_args()

    os.makedirs(args.outdir, exist_ok=True)

    df = load_and_validate(args.xlsx_path)
    print("Loaded:", df.shape)
    print("Columns:", df.columns.tolist())

    X_raw, X, _, scaler = preprocess(df)
    print("Samples:", X.shape[0], "Features:", X.shape[1])

    # Evaluate k = 2..8 with a fixed seed
    k_metrics = pd.DataFrame([eval_k(X, k, seed=args.seed_fixed) for k in range(2, 9)])
    k_metrics.to_csv(os.path.join(args.outdir, "K_comparison_metrics_seed_fixed.csv"), index=False)
    print("\nK comparison metrics saved.")

    # Tiny cluster scan across seeds
    _, tiny_summary = tiny_cluster_seed_scan(X, args.outdir, ks=(3, 4), seeds=range(50))
    print("\nTiny cluster seed scan summary:")
    print(tiny_summary)

    # Subgroup split entropy
    sub_df = compute_subgroup_split(X_raw, X, args.outdir, ks=(3, 4), seed=args.seed_fixed)
    print("\nSubgroup split results saved:")
    print(sub_df)

    # Centers (original scale) for k=3 and k=4
    c3 = centers_df(X, scaler, k=3, seed=args.seed_fixed)
    c4 = centers_df(X, scaler, k=4, seed=args.seed_fixed)

    c3.to_csv(os.path.join(args.outdir, "centers_K3_original_scale.csv"), index=False)
    c4.to_csv(os.path.join(args.outdir, "centers_K4_original_scale.csv"), index=False)

    c3[["cluster", "age", "TT", "CRP", "Cre"]].to_csv(
        os.path.join(args.outdir, "centers_K3_subset_age_TT_CRP_Cre.csv"), index=False
    )
    c4[["cluster", "age", "TT", "CRP", "Cre"]].to_csv(
        os.path.join(args.outdir, "centers_K4_subset_age_TT_CRP_Cre.csv"), index=False
    )

    print("\nCluster centers saved.")
    print(f"\nDone. Outputs written to: {args.outdir}")

if __name__ == "__main__":
    main()
